frontend/src/utils/account-send.ts

```diff
@@ -0,0 +1,179 @@
+  viewingKey: string;
+};
+
+// Encrypted values that are stored in local storage
+type EncryptedAccountSendData = {
+  encryptedAddress: string;
+};
+
+// Unencrypted values that are stored in local storage
+type UnencryptedAcountSendData = {
+  amount: string;
+  tokenAddress: string;
+  dateSent: Date;
+  txHash: string;
+  userAddress: string;
+  recipientAddress: string;
+};
+
+type EncryptAccountDataArgs = KeyData & AccountDataToEncrypt;
+type DecryptAccountDataArgs = KeyData & EncryptedAccountSendData;
+type StoreSendArgs = { chainId: number; viewingKey: string; provider: Web3Provider } & Omit<
+  AccountSendData,
+  'dateSent' | 'recipientId'
+> &
+  Omit<AccountDataToEncrypt, 'address'>;
+type FetchAccountSendArgs = {
+  chainId: number;
+  address: string;
+  viewingKey: string;
+};
+// All values stored in local storage
+type AccountSendDataWithEncryptedFields = UnencryptedAcountSendData & EncryptedAccountSendData;
+// All values in local storage with encrypted values decrypted
+type AccountSendData = UnencryptedAcountSendData & { recipientId: string } & Omit<AccountDataToEncrypt, 'address'>;
+
+export const buildAccountDataForEncryption = ({
+  address,
+  advancedMode,
+  pubKey,
+  usePublicKeyChecked,
+}: AccountDataToEncrypt) => {
+  if (!isAddress(address)) {
+    throw new Error('Invalid address');
+  }
+  if (pubKey.slice(0, 4) !== '0x04') {
+    throw new Error('Invalid public key');
+  }
+
+  address = getAddress(address).slice(2); // slice off the `0x` prefix.
+  const advancedModeHalfByte = advancedMode ? '1' : '0';
+  const usePublicKeyCheckedHalfByte = usePublicKeyChecked ? '1' : '0';
+  const remainingBytes = 11; // 20 bytes for address, 1 byte total for advancedMode and usePublicKeyChecked.
+
+  // Store as much of the pubkey as we can in the remaining bytes. The first 4 characters are `0x04` so we skip those.
+  const pubKeyStart = pubKey.slice(4, remainingBytes * 2 + 4); // Each slot is a half byte meaning remaining bytes must be multiplied by 2
+  const dataToEncrypt = `${address}${advancedModeHalfByte}${usePublicKeyCheckedHalfByte}${pubKeyStart}`;
+  if (dataToEncrypt.length !== 64) {
+    // This should never happen.
+    throw new Error(`Data to encrypt is not the correct length ${dataToEncrypt.length} instead of 64`);
+  }
+
+  return BigNumber.from(`0x${dataToEncrypt}`);
+};
+
+export const encryptAccountData = ({
+  address,
+  advancedMode,
+  usePublicKeyChecked,
+  encryptionCount,
+  viewingKey,
+  pubKey,
+}: EncryptAccountDataArgs) => {
+  const key = keccak256(toUtf8Bytes(`${viewingKey}${encryptionCount}`));
+  const data = buildAccountDataForEncryption({ address, advancedMode, usePublicKeyChecked, pubKey });
+  const encryptedData = data.xor(key);
+  return hexZeroPad(encryptedData.toHexString(), 32);
+};
+
+export const decryptData = ({ viewingKey, encryptionCount, encryptedAddress }: DecryptAccountDataArgs) => {
+  const key = keccak256(toUtf8Bytes(`${viewingKey}${encryptionCount}`));
```

Comment:
I am getting the below error when I remove toUtf8Bytes.

```
  hex data is odd-length (argument="value", value="0x29
0a15e2b46811c84a0c26624fd7fdc12e38143ae75518fc48375d41035
ec5c11", code=INVALID_ARGUMENT, version=bytes/5.7.0)
```

Reply:
This is because the hex string is 65 digits (ignoring the 0x prefix) which is 32.5 bytes, and data should be an integer number of bytes. The reason for this issue is because we concatenate a 64 char _hex string_ (the 32 byte viewing key) with a _decimal string_ (`keyData.encryptionCount`) value. We shouldn't be concatenating two different types, and `toUtf8Bytes` was basically obscuring this bug. 

The fix is converting `keyData.encryptionCount` to a 32 byte hex string then concatenating the two values, and then hashing that 64 byte string. (untested example code below, needs to be verified—it's a bit confusing because `keyData.viewingKey` has the 0x prefix, and we want it there, but `encryptionCount` also has the 0x prefix and we need to slice it off there)

```typescript
const encryptionCount = hexZeroPad(BigNumber.from(keyData.encryptionCount).toHexString(), 32);
const key = keccak256(`${keyData.viewingKey}${encryptionCount.slice(2)}`);
```

Reply:
Note that we don't have to pad `encryptionCount` to 32 bytes, we can pad to any size we want, as long as that size is larger than the highest expected count. So in theory we could pad to smaller than 32 bytes, but it's pretty much free to use 32 bytes, so may as well.

Reply:
Removed toUtfBytes here https://github.com/ScopeLift/umbra-protocol/pull/494/commits/6f34ff1725f4036c2567289011a57ebf3566c03d